### PR TITLE
[SMTChecker] Support to external calls to unknown code

### LINKED ISSUE #8972: [SMTChecker] CHC support to external calls to unknown code

### PR DESCRIPTION
**Pull request by leonardoalt**
Fixes https://github.com/ethereum/solidity/issues/8972
Depends on https://github.com/ethereum/solidity/pull/9159

tl;dr
Support reentrancy from unknown code without losing all knowledge.

This PR extends the CHC model to support external calls to unknown code. Since the code is unknown, it is sound to assume that it can nondeterministically call the analyzed contract, any number of times.
Because of that, the model is extended with 2 new types of rules that encode this nondeterminism, using a new predicate `nondet_interface(state, state')` for each analyzed contract, that allow the state to change nondeterministically from `state` to `state'`.

The first rule is a fact added for each contract:
1. `nondet_interface(state, state)`
This fact is the base case of the inductive rule, and allows the nondeterministic step to be applied without state changes.

The second rule is the inductive rule over each public function `f` of the contract:
2. `nondet_interface(state, state') && summary_f(state', state'') => nondet_interface(state, state'')`

When an external call to unknown code is seen, CHC simply adds the constraint
`nondet_interface(currentState, postState)` to the current block/rule. This means that the current contract can be called nondeterministically any number of times (including zero) before moving on to the next block.
Note that because it is nondeterministic, if a certain path allows for state changes in the contract, it will necessarily be reached and taken into account, so this model is sound.

If the contract has public functions that allow arbitrary state changes, knowledge will be lost at this point, but there's not much else one can do.
However, if the contract has a certain behavior where the current state might limit state changes in reentrant calls, this will be taken into account when generating invariants.
See the `mutex.sol` example added in this PR.

### CODE REVIEW

**Review by ekpyron — COMMENTED** [review]
For the record: this needs a rebase ;-), but this looks quite good to me! I'm actually almost approving it right away, but maybe I should take some more time trying to come up with something that might break it...

One remark though: we could actually check if the function called is ``view`` or ``pure``, i.e. if it'll be called with ``STATICCALL`` - in those cases this is actually too strong, i.e. we could in fact just keep the state for such calls (but we can also just relax those cases later, but it shouldn't actually be too complicated to do it here).

**Comment by leonardoalt** [comment]
Right, that makes sense! (re `staticcall`).

Yea, I also have a few test cases in mind that I still want to add.

**Comment by leonardoalt** [comment]
Rebased on top of https://github.com/ethereum/solidity/pull/9176 to see if the nondet problem is gone

**Comment by leonardoalt** [comment]
Apparently the nondeterminism in the tests in my local arch build vs CI was that the arch z3 uses gmp whereas the others don't.

**Comment by leonardoalt** [comment]
Rebased, ready for review.

**Comment by leonardoalt** [comment]
https://github.com/ethereum/solidity/pull/9159 helps with the nondeterminism

File: test/libsolidity/smtCheckerTests/external_calls/external_hash_known_code_state_reentrancy.sol [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+		assert(z == y);
+		assert(prevOwner == owner);
+	}
+
+	function g() public view returns (uint) {
+		return y;
+	}
+}
+// ----
+// Warning: (306-320): Assertion violation happens here
```
Comment (ekpyron) [inline test/libsolidity/smtCheckerTests/external_calls/external_hash_known_code_state_reentrancy.sol]: Which one is this?

Comment (leonardoalt) [reply test/libsolidity/smtCheckerTests/external_calls/external_hash_known_code_state_reentrancy.sol]: z == y

Comment (leonardoalt) [reply test/libsolidity/smtCheckerTests/external_calls/external_hash_known_code_state_reentrancy.sol]: it fails because `State.c` is not necessarily the caller

Comment (leonardoalt) [reply test/libsolidity/smtCheckerTests/external_calls/external_hash_known_code_state_reentrancy.sol]: and since the state is not carried over to the called contract, it doesn't know anything about `c`'s state

File: libsolidity/formal/CHC.cpp [hunk 2]
```diff
@@ -574,6 +580,33 @@ void CHC::internalFunctionCall(FunctionCall const& _funCall)
...
+	FunctionType const& funType = dynamic_cast<FunctionType const&>(*_funCall.expression().annotation().type);
+	auto kind = funType.kind();
+	solAssert(kind == FunctionType::Kind::External || kind == FunctionType::Kind::BareStaticCall, "");
+
+	auto const* function = functionCallToDefinition(_funCall);
+	if (!function)
+		return;
+
+	auto preCallState = currentStateVariables();
+	bool noStateChanges = kind == FunctionType::Kind::BareStaticCall ||
```
Comment (ekpyron) [inline libsolidity/formal/CHC.cpp]: I might have gone for something like
```suggestion
	bool usesStaticCall = kind == FunctionType::Kind::BareStaticCall ||
```
to avoid the "if not no state changes" below...

**Comment by Marenz** [comment]
This needs a rebase to resolve the conflicts

**Review by ekpyron — DISMISSED** [review]

File: libsolidity/formal/CHC.cpp (hunk 2 above)
Comment (leonardoalt) [reply libsolidity/formal/CHC.cpp]: Updated

**Review by ekpyron — APPROVED** [review]

**Review by ekpyron — CHANGES_REQUESTED** [review]
We decided to change the implemented pure function case.

**Comment by leonardoalt** [comment]
@ekpyron updated

**Review by ekpyron — APPROVED** [review]
I wonder what's the best way to have CHC track if we know that the address in a state variable like ``C c;`` is of a contract type ``C`` for which we know the code (i.e. CHC sees where it is deployed)...

Maybe there's just a boolean flag for ``C c;`` that means "contains valid ``C``" and that is initialized to ``false`` and set to ``true`` on any ``c = new C()``? And then calling ``c.f()`` will take the undefined path, if it's false and assume the code of ``C.f`` being called, if it's ``true``?

But yeah - maybe we can really just assume that everything named ``C`` is actually a ``C``, I'm not entirely sure... it would feel weird to me in any case.

**Comment by leonardoalt** [comment]
@ekpyron created https://github.com/ethereum/solidity/issues/9287 to track that
